[Feature] Add features to extend expressions functionalities

Doomsk opened the pull request:
This PR aimed to accomplish the following

- [x] (**NO LONGER VALID**) extend arithmetic operations between `Expression` and numeric values, such as int, float, complex, to also external packages such as `numpy` and `torch` (currently only native python numeric types supported)
- [x] remove `withrepr` decorator. It caused issues for not supporting kwargs, such as the case of `NativeDrive` with its `instruction_name` kwarg
- [x] include `__matmul__` (`@`) operation as a kron operation on `Expression`

kaosmicadei commented on qadence2_expressions/core/expression.py:
```diff
@@ -558,11 +559,14 @@ def __kron__(self, other: object) -> Expression:
+        # forces to add the evaluation at this point.
         return evaluate_kron(Expression.kron(self, other))
 
+    def __matmul__(self, other: object) -> Expression:
+        return self.__kron__(other)
```
The reason it wasn't exposed initially was to avoid confusion on how multiplication works.

Having both options, `X(1) * Z(2)` and `X(1) @ Z(2)`, adds an unnecessary level of complexity from the use point of view.

RolandMacDoland replied:
I would also think so. Maybe this is needed for backwards compatibility purposes ?

kaosmicadei commented on qadence2_expressions/core/utils.py:
```diff
@@ -0,0 +1,32 @@
+@runtime_checkable
+class Numeric(Protocol):
+    """
+    Protocol class to represent Python base classes for numeric values,
+    such as complex, float, int, but also include packages types, such
+    as numpy and torch. It is especially useful for handling operations
+    with Expressions and any kind of "Numeric" type.
```
Since the expressions only handle Python inputs, what is the point of adding support to `numpy` or `torch` at this point?

Doomsk replied:
to be backwards compatible

kaosmicadei commented:
I'm not sure if Pyhton allows operator alias. If not, then a `DeprecationWarning` should be added to the `__matmul__` definition to emphasise the backward compatibility. About the `Numeric`, I agree, @RolandMacDoland, as long as the mention of `numpy` and `torch` is removed.
> ~but also include packages types, such as numpy and torch.~

Doomsk commented:
@RolandMacDoland @kaosmicadei 

- [x] removed: `Numeric`
- [x] removed: `withrepr`
- [ ] removed: `__matmul__`; caveats: no backwards compatibility since the method can only exists in the `Expression` class; a `kron` function on legacy already exists

RolandMacDoland commented on qadence2_expressions/core/constructors.py:
```diff
@@ -63,7 +33,8 @@ def promote(x: Expression | complex | float | int) -> Expression:
 
     Args:
-        x (Expression | complex | float | int): A valid expression or numerical value. Numerical values are converted into `Value(x)` expressions.
+        x (Expression | complex | float | int): A valid expression or numerical value.
```
I guess the bottomline is: `Numeric` type is good as it makes `Expression` more readable than to carry over this long union type everywhere.

pimvenderbosch replied:
I'm with @RolandMacDoland that the `Numeric` type is a good addition.

Doomsk replied:
The problem with using `Numeric` protocol as before is that there will be the need to implement some exception handlings throughout the `Expression` class. For instance:

```
from qadence2_expressions import Numeric
import torch
isinstance(torch.tensor(1.0, requires_grad=True), Numeric)

>>
Traceback (most recent call last):
  File "/Users/emaschio/Library/Application Support/hatch/env/virtual/qadence2-expressions/WGNTQKyX/qadence2-expressions/lib/python3.10/site-packages/IPython/core/interactiveshell.py", line 3577, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-72c52bf21fe0>", line 4, in <module>
    isinstance(torch.tensor(1.0, requires_grad=True), Numeric)
  File "/opt/homebrew/Cellar/python@3.10/3.10.15/Frameworks/Python.framework/Versions/3.10/lib/python3.10/typing.py", line 1506, in __instancecheck__
    if all(hasattr(instance, attr) and
  File "/opt/homebrew/Cellar/python@3.10/3.10.15/Frameworks/Python.framework/Versions/3.10/lib/python3.10/typing.py", line 1506, in <genexpr>
    if all(hasattr(instance, attr) and
RuntimeError: imag is not implemented for tensors with non-complex dtypes.
```

while 

```
from qadence2_expressions import Numeric
import torch
isinstance(torch.tensor(1.0j, requires_grad=True), Numeric)

>> True
```

Doomsk replied:
nevermind. Replaced it by `Numeric = Union[complex | float | int]` alias

RolandMacDoland commented on qadence2_expressions/core/expression.py:
```diff
@@ -489,6 +490,18 @@ def __pow__(self, other: object) -> Expression:
 
+        # Power of power is a simple operation and can be evaluated here.
+        if (
+            self.is_quantum_operator
```
Hm there is a number this checks sometimes. Just wondering whether they couldn't be grouped under fewer (possibly single) predicate(s).

Doomsk replied:
can you point where you found them out? Probably we could group them into a single function

RolandMacDoland commented on tests/test_expression.py:
```diff
@@ -93,16 +93,20 @@ def test_kron() -> None:
 
     # Push term from the right.
     assert term.__kron__(X(3)) == Expression.kron(X(1), X(2), X(3), X(4))
+    assert term @ X(3) == Expression.kron(X(1), X(2), X(3), X(4))
```
Maybe check the warning has been raised ?

RolandMacDoland reviewed (commented):
Thanks @Doomsk for this. I guess you can reintroduce the `Numeric` type and I'd be happy to approve soon after. Will seek for @pimvenderbosch's review anyway.

pimvenderbosch reviewed (commented):
Hey @Doomsk, thanks for the work. Looks good in general, minor comments.

Doomsk commented on pyproject.toml:
```diff
@@ -6,7 +6,7 @@ build-backend = "hatchling.build"
 description = "Qadence 2 expressions consisting of symbolic expressions and blocks, i.e. quantum digital, analog, digital-analog and composite gates, together with their logic and engine."
 readme = "README.md"
-version = "0.1.3"
+version = "0.2.0"
```
better to have a PR just for the bump indeed

pimvenderbosch reviewed (approved):
Thanks @Doomsk! Ready for merge imo
